File: scripts/slugify.py
"""
Python port of scripts/slugify.js — kept logically identical (same regexes,
same core-name extraction) since fetch_feed.py (Python) is what actually
persists the "slug" field in products.json; generate_static_pages.js /
generate_merchant_feed.js (Node) just read p.slug back out, they never
recompute it. See scripts/slugify.js for the full rationale.
"""

import re

TRANSLIT = {
    'а': 'a', 'б': 'b', 'в': 'v', 'г': 'h', 'ґ': 'g', 'д': 'd', 'е': 'e', 'є': 'ie', 'ж': 'zh',
    'з': 'z', 'и': 'y', 'і': 'i', 'ї': 'i', 'й': 'i', 'к': 'k', 'л': 'l', 'м': 'm', 'н': 'n',
    'о': 'o', 'п': 'p', 'р': 'r', 'с': 's', 'т': 't', 'у': 'u', 'ф': 'f', 'х': 'kh', 'ц': 'ts',
    'ч': 'ch', 'ш': 'sh', 'щ': 'shch', 'ь': '', 'ю': 'iu', 'я': 'ia', "'": '',
    'ё': 'e', 'э': 'e', 'ы': 'y', 'ъ': '',
}

TYPE_PREFIXES = [
    r'^Комплект силових кабелів\s*',
    r'^Сонячний кабель\s*',
    r'^Гібридний інвертор\s*',
    r'^Акумуляторна батарея\s*',
    r'^Акумулятор\s*',
    r'^Зарядна станція\s*',
    r'^Безперебійник(?:\s+для\s+\S+)?\s*',
]

UNIT_CUT_RE = re.compile(
    r'\d+(?:[.,]\d+)?\s*(?:кВт|kW|kw|Вт|фаза|V|В|Ah|Аh|ah)(?![A-Za-zА-Яа-яІіЇїЄєҐґ])', re.I
)
WIFI_RE = re.compile(r'\bWi-?Fi\b', re.I)


def strip_type_prefix(name):
    for pat in TYPE_PREFIXES:
        stripped = re.sub(pat, '', name, count=1, flags=re.I)
        if stripped != name:
            return stripped.strip()
    return name


def part_short(segment):
    cuts = []
    m1 = UNIT_CUT_RE.search(segment)
    if m1:
        cuts.append(m1.start())
    m2 = WIFI_RE.search(segment)
    if m2:
        cuts.append(m2.start())
    s = segment[:min(cuts)] if cuts else segment
    return s.strip().rstrip(',+').strip()


def kit_core_name(title):
    body = re.sub(r'^Комплект автономного енергоживлення:\s*', '', title, flags=re.I)
    parts = re.split(r'\s*\+\s*', body, maxsplit=1)
    if len(parts) != 2:
        return part_short(body)
    return part_short(parts[0]) + ' ' + part_short(parts[1])


def core_name(title):
    if re.match(r'^Комплект автономного енергоживлення', title, re.I):
        return kit_core_name(title)
    return part_short(strip_type_prefix(title))


def transliterate(s):
    return ''.join(TRANSLIT.get(ch.lower(), ch) for ch in s)


def truncate_slug(slug, max_len):
    if len(slug) <= max_len:
        return slug
    cut = slug[:max_len]
    if '-' in cut:
        cut = cut.rsplit('-', 1)[0]
    return cut


def slugify(title, max_len=65):
    name = core_name(title or '')
    translit = transliterate(name).lower()
    slug = re.sub(r'[^a-z0-9]+', '-', translit)
    slug = re.sub(r'-{2,}', '-', slug).strip('-')
    return truncate_slug(slug, max_len)

File: scripts/slugify.js
/**
 * Shared slug generator for product URLs: /product/{slug}/{id}.html
 *
 * Slugs are meant to be computed ONCE per product id and then persisted in
 * products.json ("slug" field) — see fetch_feed.py's sticky carry-forward
 * for description/link/specs, which slug generation follows the same
 * pattern for. Nothing here should be called at request time; it's a
 * build-time helper only.
 *
 * Core-name extraction mirrors miniapp_v/index.html's stripTypePrefix() /
 * kitShortName() / kitPartShort() — same idea, same regexes, ported here so
 * slugs read as "brand + model (+ 2nd component for kits)" instead of the
 * full sentence-like product title (which repeats voltage/capacity/weight
 * that don't belong in a URL).
 */

const TRANSLIT = {
  а: 'a', б: 'b', в: 'v', г: 'h', ґ: 'g', д: 'd', е: 'e', є: 'ie', ж: 'zh',
  з: 'z', и: 'y', і: 'i', ї: 'i', й: 'i', к: 'k', л: 'l', м: 'm', н: 'n',
  о: 'o', п: 'p', р: 'r', с: 's', т: 't', у: 'u', ф: 'f', х: 'kh', ц: 'ts',
  ч: 'ch', ш: 'sh', щ: 'shch', ь: '', ю: 'iu', я: 'ia', "'": '',
  ё: 'e', э: 'e', ы: 'y', ъ: '',
};

const TYPE_PREFIXES = [
  /^Комплект силових кабелів\s*/i,
  /^Сонячний кабель\s*/i,
  /^Гібридний інвертор\s*/i,
  /^Акумуляторна батарея\s*/i,
  /^Акумулятор\s*/i,
  /^Зарядна станція\s*/i,
  /^Безперебійник(?:\s+для\s+\S+)?\s*/i,
];

// Same as miniapp_v/index.html's KIT_UNIT_CUT_RE / WIFI_RE.
const UNIT_CUT_RE = /\d+(?:[.,]\d+)?\s*(?:кВт|kW|kw|Вт|фаза|V|В|Ah|Аh|ah)(?![A-Za-zА-Яа-яІіЇїЄєҐґ])/i;
const WIFI_RE = /\bWi-?Fi\b/i;

function stripTypePrefix(name) {
  for (const re of TYPE_PREFIXES) {
    const stripped = name.replace(re, '');
    if (stripped !== name) return stripped.trim();
  }
  return name;
}

function partShort(segment) {
  const cuts = [];
  const m1 = segment.match(UNIT_CUT_RE);
  if (m1) cuts.push(m1.index);
  const m2 = segment.match(WIFI_RE);
  if (m2) cuts.push(m2.index);
  const s = cuts.length ? segment.slice(0, Math.min(...cuts)) : segment;
  return s.trim().replace(/[,+]+$/, '').trim();
}

/* "Комплект автономного енергоживлення: X + Y" -> "X Y" (short codes only). */
function kitCoreName(title) {
  const body = title.replace(/^Комплект автономного енергоживлення:\s*/i, '');
  const parts = body.split(/\s*\+\s*/);
  if (parts.length !== 2) return partShort(body);
  return partShort(parts[0]) + ' ' + partShort(parts[1]);
}

function coreName(title) {
  if (/^Комплект автономного енергоживлення/i.test(title)) return kitCoreName(title);
  return partShort(stripTypePrefix(title));
}

function transliterate(str) {
  let out = '';
  for (const ch of str) {
    const lower = ch.toLowerCase();
    out += Object.prototype.hasOwnProperty.call(TRANSLIT, lower) ? TRANSLIT[lower] : ch;
  }
  return out;
}

/* Truncate on a '-' boundary so we never cut a word in half. */
function truncateSlug(slug, maxLen) {
  if (slug.length <= maxLen) return slug;
  let cut = slug.slice(0, maxLen);
  if (cut.includes('-')) cut = cut.slice(0, cut.lastIndexOf('-'));
  return cut;
}

function slugify(title, maxLen = 65) {
  const name = coreName(title || '');
  const translit = transliterate(name).toLowerCase();
  let slug = translit.replace(/[^a-z0-9]+/g, '-').replace(/-{2,}/g, '-').replace(/^-+|-+$/g, '');
  return truncateSlug(slug, maxLen);
}

module.exports = { slugify };
